### PR TITLE
Disallow field indirection in INSERT/UPDATE queries

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Disallows field indirection in INSERT/UPDATE queries

As described in https://github.com/citusdata/citus/issues/1293, we were not handling field indirection in INSERT/UPDATE queries. 
Changes in https://github.com/citusdata/citus/pull/4234 attempt to fix this issue but not sufficient and requires more scoping (https://github.com/citusdata/citus/pull/4234#issuecomment-707658066).
So for now, we disallow field indirection.

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -674,6 +674,15 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
 			{
 				Assert(hasVarArgument || hasBadCoalesce);
 			}
+
+			if (IsA((Node *) targetEntry->expr, FieldStore))
+			{
+				/* DELETE cannot do field indirection already */
+				Assert(commandType == CMD_UPDATE || commandType == CMD_INSERT);
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "inserting or modifying composite type fields is not "
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:683]: can we give a hint on how to do  it?

Comment (onurctirtir) [reply src/backend/distributed/planner/multi_router_planner.c:683]: I will give a hint to prefer insert/update the whole composite type

Comment (onurctirtir) [reply src/backend/distributed/planner/multi_router_planner.c:683]: Does it look good ?
```c
				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
									 "inserting or modifying composite type fields is not "
									 "supported", NULL,
									 "To workaround this, use whole column for inserting or "
									 "modifying the composite type fields");
```

**Comment by onurctirtir** [comment]
Currently this pr can't catch field indirection to domain composite types:

```sql
create type insert_test_type as (if1 int, if2 int);
CREATE DOMAIN dom AS insert_test_type CHECK ((VALUE).if1 > 0);
SELECT run_command_on_workers(
$$
    CREATE DOMAIN dom AS insert_test_type CHECK ((VALUE).if1 > 0);
$$);
create table inserttest (f1 int, f3 dom);
SELECT create_distributed_table('inserttest','f1');
```

this crashes, `insert into inserttest (f1,f3.if1, f3.if2) values (0,1,2);`  
this doesn't  `insert into inserttest (f1,f3.if1) values (0,1);`

indeed, both should error out

File: src/backend/distributed/planner/multi_router_planner.c (hunk 1 above)
Comment (marcocitus) [reply src/backend/distributed/planner/multi_router_planner.c:685]: Maybe "Use the column name to insert or update the composite type as a single value"

File: src/test/regress/expected/distributed_types.out [hunk 2]
```diff
@@ -425,7 +425,7 @@ $$);
  (localhost,57638,t,"CREATE DOMAIN")
 (2 rows)
 
-CREATE TABLE domain_indirection_test (f1 int, f3 domain);
+CREATE TABLE domain_indirection_test (f1 int, f3 domain, domain_array domain[]);
```
Comment (onurctirtir) [inline src/test/regress/expected/distributed_types.out:428]: For readers, I couldn't add this column with `ALTER TABLE` command due to  https://github.com/citusdata/citus/issues/3678
That is why I changed `CREATE TABLE` command

File: src/backend/distributed/planner/multi_router_planner.c [hunk 3]
```diff
@@ -674,6 +676,17 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
 			{
 				Assert(hasVarArgument || hasBadCoalesce);
 			}
+
+			if (ExprHasFieldStore(targetEntry->expr))
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c:680]: We have a generic function for finding nodes called `FindNodeMatchingCheckFunction`, which takes as a parameter a function that returns true if the node matches (e.g. IsA FieldStore).

**Review by marcocitus — APPROVED** [review]

**Pull request merged**